drivers/sdhc/sam_hsmci.c

```diff
@@ -0,0 +1,673 @@
+
+	if (config->carrier_detect.port == NULL) {
+		return 1;
+	}
+
+	return gpio_pin_get_dt(&config->carrier_detect);
+}
+
+static int sam_hsmci_card_busy(const struct device *dev)
+{
+	const struct sam_hsmci_config *config = dev->config;
+	Hsmci *hsmci = config->base;
+
+	return (hsmci->HSMCI_SR & HSMCI_SR_NOTBUSY) == 0;
+}
+
+static void sam_hsmci_send_clocks(Hsmci *hsmci)
+{
+	hsmci->HSMCI_MR &= ~(HSMCI_MR_WRPROOF | HSMCI_MR_RDPROOF | HSMCI_MR_FBYTE);
+	hsmci->HSMCI_ARGR = 0;
+	hsmci->HSMCI_CMDR =
+		HSMCI_CMDR_RSPTYP_NORESP | HSMCI_CMDR_SPCMD_INIT | HSMCI_CMDR_OPDCMD_OPENDRAIN;
+	while (!(hsmci->HSMCI_SR & HSMCI_SR_CMDRDY)) {
+		;
+	}
+	hsmci->HSMCI_MR |= HSMCI_MR_WRPROOF | HSMCI_MR_RDPROOF;
+}
+
+static int sam_hsmci_send_cmd(Hsmci *hsmci, struct sdhc_command *cmd, uint32_t cmdr,
+			      struct sam_hsmci_data *data)
+{
+	uint32_t sr;
+
+	hsmci->HSMCI_ARGR = cmd->arg;
+
+	cmdr |= HSMCI_CMDR_CMDNB(cmd->opcode) | HSMCI_CMDR_MAXLAT_64;
+	if (data->open_drain) {
+		cmdr |= HSMCI_CMDR_OPDCMD_OPENDRAIN;
+	}
+
+	uint8_t nrt = cmd->response_type & SDHC_NATIVE_RESPONSE_MASK;
+
+	if (nrt > SD_RSP_TYPE_R7) {
+		return -ENOTSUP;
+	}
+
+	cmdr |= _resp2size[nrt];
+	hsmci->HSMCI_CMDR = cmdr;
+	do {
+		sr = hsmci->HSMCI_SR;
+
+		/* special case ,ignore CRC status if response is R3 to clear it */
+		if (nrt == SD_RSP_TYPE_R3 || nrt == SD_RSP_TYPE_NONE) {
+			sr &= ~HSMCI_SR_RCRCE;
+		}
+
+		if ((sr & _HSMCI_SR_ERR) != 0) {
+			LOG_DBG("Status register error bits: %08x", sr & _HSMCI_SR_ERR);
+			return -EIO;
+		}
+	} while (!(sr & HSMCI_SR_CMDRDY));
+
+	if (nrt == SD_RSP_TYPE_R1b) {
+		do {
+			sr = hsmci->HSMCI_SR;
+		} while (!((sr & HSMCI_SR_NOTBUSY) && ((sr & HSMCI_SR_DTIP) == 0)));
+	}
+
+	/* RSPR is just a FIFO, index is of no consequence */
+	cmd->response[3] = hsmci->HSMCI_RSPR[0];
+	cmd->response[2] = hsmci->HSMCI_RSPR[0];
+	cmd->response[1] = hsmci->HSMCI_RSPR[0];
+	cmd->response[0] = hsmci->HSMCI_RSPR[0];
+	return 0;
+}
+
+static int sam_hsmci_wait_write_end(Hsmci *hsmci)
+{
+	uint32_t sr = 0;
+
+	/* Timeout is included in HSMCI, see DTOE bit, not required explicitly. */
+	do {
+		sr = hsmci->HSMCI_SR;
+		if (sr & (HSMCI_SR_UNRE | HSMCI_SR_OVRE | HSMCI_SR_DTOE | HSMCI_SR_DCRCE)) {
+			LOG_DBG("PDC sr 0x%08x error", sr);
+			return -EIO;
+		}
+	} while (!(sr & HSMCI_SR_TXBUFE));
+
+	do {
+		sr = hsmci->HSMCI_SR;
+		if (sr & (HSMCI_SR_UNRE | HSMCI_SR_OVRE | HSMCI_SR_DTOE | HSMCI_SR_DCRCE)) {
+			LOG_DBG("PDC sr 0x%08x last transfer error", sr);
+			return -EIO;
+		}
+	} while (!(sr & HSMCI_SR_NOTBUSY));
+
+	if (!(hsmci->HSMCI_SR & HSMCI_SR_FIFOEMPTY)) {
+		return -EIO;
+	}
+	return 0;
+}
+
+static int sam_hsmci_wait_read_end(Hsmci *hsmci)
+{
+	uint32_t sr;
+
+	do {
+		sr = hsmci->HSMCI_SR;
+		if (sr & (HSMCI_SR_UNRE | HSMCI_SR_OVRE | HSMCI_SR_DTOE | HSMCI_SR_DCRCE)) {
+			LOG_DBG("PDC sr 0x%08x error", sr & (HSMCI_SR_UNRE | HSMCI_SR_OVRE |
+							     HSMCI_SR_DTOE | HSMCI_SR_DCRCE));
+			return -EIO;
+		}
+	} while (!(sr & HSMCI_SR_RXBUFF));
```

Comment:
I dropped all this **} while (!(sr & HSMCI_SR_RXBUFF));** block to build for SAMV71 and SAM4E still passing on tests.
Can you check if this is really necessary?

Reply:
You are right. These registers are for PDC, which the V71 does not have. The SAM4E should use it when using PDC. I have wrapped these functions in ifdefs. I'll hope to update the PR today.

Reply:
@nandojve It works from my side. Should work on the V71 also. Can you check please?